pkg/operation/botanist/addons.go

```diff
@@ -443,8 +409,105 @@ func (b *Botanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart, erro
 		"node-problem-detector": common.GenerateAddonConfig(nodeProblemDetector, true),
 		"podsecuritypolicies":   common.GenerateAddonConfig(podSecurityPolicies, true),
 		"shoot-info":            common.GenerateAddonConfig(shootInfo, true),
-		"vpn-shoot":             common.GenerateAddonConfig(vpnShoot, true),
-	})
+	}
+
+	// If kubernetes version is >= 1.18 deploy the proxy-agent, otherwise deploy the proxy server
+	kubernetesVersionGeq118, err := versionutils.CheckVersionMeetsConstraint(b.Shoot.Info.Spec.Kubernetes.Version, ">= 1.18")
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		tunnelSecret            *corev1.Secret
+		konnectivityAgentConfig = map[string]interface{}{
+			"proxyHost": common.GetAPIServerDomain(b.Shoot.InternalClusterDomain),
+			"podAnnotations": map[string]interface{}{
+				"checksum/secret-konnectivity-agent": b.CheckSums["konnectivity-agent"],
+			},
+			"kubernetesVersion": b.Shoot.Info.Spec.Kubernetes.Version,
+		}
+		vpnTLSAuthSecret = b.Secrets["vpn-seed-tlsauth"]
+		vpnShootConfig   = map[string]interface{}{
+			"podNetwork":     b.Shoot.Networks.Pods.String(),
+			"serviceNetwork": b.Shoot.Networks.Services.String(),
+			"tlsAuth":        vpnTLSAuthSecret.Data["vpn.tlsauth"],
+			"podAnnotations": map[string]interface{}{
+				"checksum/secret-vpn-shoot": b.CheckSums["vpn-shoot"],
+			},
+			"kubernetesVersion": b.Shoot.Info.Spec.Kubernetes.Version,
+		}
+		konnectivityAgentSecret = b.Secrets["konnectivity-agent"]
+		vpnShootSecret          = b.Secrets["vpn-shoot"]
+
+		shootClient = b.K8sShootClient.Client()
+	)
+
+	// Konnectivity agent related values
+	konnectivityAgent, err := b.InjectShootShootImages(konnectivityAgentConfig, common.KonnectivityAgentImageName)
+	if err != nil {
+		return nil, err
+	}
+	values["konnectivity-agent"] = common.GenerateAddonConfig(konnectivityAgent, true)
+
+	// OpenVPN related values
+	if openvpnDiffieHellmanSecret, ok := b.Secrets[common.GardenRoleOpenVPNDiffieHellman]; ok {
+		vpnShootConfig["diffieHellmanKey"] = openvpnDiffieHellmanSecret.Data["dh2048.pem"]
+	}
```

Comment:
Why are we doing all these things here unconditionally, i.e. we don't check if Kubernetes >= 1.18? It doesn't harm too much, but it would be more clear if you would only compute the VPN-Shoot config if < 1.18, and the konnectivity only if >= 1.18.

Reply:
we still need some values to be set in either case, (e.g., kubernetesVersion).

Reply:
this will not be needed if we introduce the extra variable.